### ISSUE TITLE
segmentation fault when using pytorch 1.1.0  num_workers>0 

### ISSUE THREAD

**Issue by weixiong-ur** [horovod/horovod#1215]
**Environment:**
1. Framework: Pytorch
2. Framework version: 1.1.0 
3. Horovod version: 0.16.4
4. MPI version: 
5. CUDA version: 10.0
6. NCCL version: 
7. Python version: 3.6
8. OS and version: ubuntu
9. GCC version: 4.9.2

I run my pytorch code with horovod. If I run  code with num_workers=0, then everything is fine. However, if I run the same code with num_workers > 0, I will get a segmentation fault. Anyone encoutered the same case?

The error messages are as follows:
1,0]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,0]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,0]<stderr>:[1,0]<stderr>:2019-07-13 05:05:31,244  noisy image PSNR: 0.0000 
[1,0]<stderr>:2019-07-13 05:05:31,389  [epoch 0] validation PSNR: 22.2650 validation l2 loss: 0.00876302 Best Validation l2 loss: 0.00876302 
[1,0]<stderr>:2019-07-13 05:05:31,390  current lr is: 0.0010000
[1,4]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,4]<stderr>:[1,5]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,5]<stderr>:[1,5]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,5]<stderr>:[1,6]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,6]<stderr>:[1,1]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,1]<stderr>:[1,1]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,1]<stderr>:[1,0]<stderr>:2019-07-13 05:05:45,208  [Epoch 1] pixel loss: 0.006785 gen_adv_loss: 0.00000000 dis_adv_loss: 0.00000000 tv_loss: 0.000000 color_loss: 0.000091 percep_loss: 0.0000 style_loss: 0.000000 PSNR: 0.0000 time elapsed: 13.82
[1,0]<stderr>:2019-07-13 05:05:45,209  current lr is: 0.0010000
[1,7]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,7]<stderr>:[1,3]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,3]<stderr>:[1,5]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,5]<stderr>:[1,3]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,3]<stderr>:[1,5]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,5]<stderr>:[1,4]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,4]<stderr>:[1,0]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,0]<stderr>:[1,2]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,2]<stderr>:[1,0]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,0]<stderr>:[1,0]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,0]<stderr>:[1,0]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,0]<stderr>:[1,6]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,6]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,6]<stderr>:[1,2]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,2]<stderr>:[1,4]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,4]<stderr>:[1,1]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,1]<stderr>:[1,1]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,1]<stderr>:[1,1]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,1]<stderr>:[1,1]<stderr>:ERROR: Unexpected segmentation fault encountered in worker.
[1,1]<stderr>:[1,0]<stderr>:2019-07-13 05:05:59,594  [Epoch 2] pixel loss: 0.006428 gen_adv_loss: 0.00000000 dis_adv_loss: 0.00000000 tv_loss: 0.000000 color_loss: 0.000091 percep_loss: 0.0000 style_loss: 0.000000 PSNR: 0.0000 time elapsed: 14.38

**Comment by weixiong-ur** [comment]
I run my code on 1 node with 8 tesla V100  gpus. I already installed horovod with pip3 install --no-cache-dir horovod. open mpi version is 4.0.0, nccl version is 2.3.7

**Comment by Annusha** [comment]
I have the same issue. With num_wrokers=0 everything is fine.
In my case I have it only during feature extraction, so I just rerun process as soon as it fails, but anyway super annoying and could happen during training. 
pytorch 1.1.0 as well

**Comment by alsrgv** [comment]
@weixiong-ur, do you have RDMA NICs, e.g. RoCE or InfiniBand?  Which MPI version are you using?

**Comment by YoungDav** [comment]
i meet the same problem

**Comment by shyhuai** [comment]
Does anyone address the problem? I met the same error after running about 2 epochs on the data set of ImageNet-2012. The configurations of my cluster are Horovod-0.18.2, PyTorch-1.1.0, Python3.5, OpenMPI-4.0.1, NCCL-2.4.7, 100Gbps InfiniBand and RTX 2080Ti.

**Comment by tgaddair** [comment]
Hey @shyhuai, thanks for the details.  Are you training ImageNet with the provided [pytorch_imagenet_resnet50.py](https://github.com/horovod/horovod/blob/v0.18.2/examples/pytorch_imagenet_resnet50.py) script or your own training script?

Can you open the stack trace in gdb and paste the stack trace here?

**Comment by shyhuai** [comment]
Hi, @tgaddair , I train ImageNet with my own scripts, but the main codes are similar to [pytorch_imagenet_resnet50.py](https://github.com/horovod/horovod/blob/v0.18.2/examples/pytorch_imagenet_resnet50.py) except the data part. Instead of using the built-in `datasets.ImageFolder`, I used my own defined dataset as [this one](https://github.com/hclhkbu/gtopkssgd/blob/master/datasets.py). It works well when I run the same training scripts on another cluster (4 nodes with 56Gbps IB and each node has 4 V100 GPUs). Could you suggest how to open the stack trace? I installed horovod with pip. Thanks!

**Comment by tgaddair** [comment]
Hey @shyhuai, first you need to setup your system to write out core dumps.  If you're using Ubuntu, there is some guidance [here](https://askubuntu.com/questions/966407/where-do-i-find-the-core-dump-in-ubuntu-16-04lts) on how to set that up.

Once you know where the core dump is, you can open it in gdb and run `bt` to get that backtrace:

```
gdb <core_file>
$ bt
```

**Comment by shyhuai** [comment]
Hi @tgaddair , the backtrace from the core dump file is as follows. Thanks!
```
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/python3.5...(no debugging symbols found)...done.

warning: core file may not match specified executable file.
[New LWP 83972]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `python3 horovod_trainer.py --dnn resnet50 --dataset imagenet --max-epochs 40 --'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fe1f8f79269 in raise (sig=11) at ../sysdeps/unix/sysv/linux/pt-raise.c:35
35      ../sysdeps/unix/sysv/linux/pt-raise.c: No such file or directory.
(gdb) bt
#0  0x00007fe1f8f79269 in raise (sig=11) at ../sysdeps/unix/sysv/linux/pt-raise.c:35
#1  0x00007fe1f47b7295 in handler_SIGSEGV(int, siginfo_t*, void*) ()
   from /home/esetstore/.local/lib/python3.5/site-packages/torch/lib/libtorch_python.so
#2  <signal handler called>
#3  _int_malloc (av=av@entry=0x7fe1f8f62b20 <main_arena>, bytes=bytes@entry=83328) at malloc.c:3612
#4  0x00007fe1f8c22184 in __GI___libc_malloc (bytes=83328) at malloc.c:2913
#5  0x00007fe1d3e66b88 in memory_get_block ()
   from /home/esetstore/.local/lib/python3.5/site-packages/PIL/_imaging.cpython-35m-x86_64-linux-gnu.so
#6  0x00007fe1d3e66d51 in ImagingAllocateArray ()
   from /home/esetstore/.local/lib/python3.5/site-packages/PIL/_imaging.cpython-35m-x86_64-linux-gnu.so
#7  0x00007fe1d3e66f2b in ImagingNewInternal ()
   from /home/esetstore/.local/lib/python3.5/site-packages/PIL/_imaging.cpython-35m-x86_64-linux-gnu.so
#8  0x00007fe1d3e321cf in ImagingResampleInner ()
   from /home/esetstore/.local/lib/python3.5/site-packages/PIL/_imaging.cpython-35m-x86_64-linux-gnu.so
#9  0x00007fe1d3e32430 in ImagingResample ()
   from /home/esetstore/.local/lib/python3.5/site-packages/PIL/_imaging.cpython-35m-x86_64-linux-gnu.so
#10 0x00007fe1d3e27e8b in ?? () from /home/esetstore/.local/lib/python3.5/site-packages/PIL/_imaging.cpython-35m-x86_64-linux-gnu.so
#11 0x00000000004ea10f in PyCFunction_Call ()
#12 0x0000000000536d94 in PyEval_EvalFrameEx ()
#13 0x000000000053fc97 in ?? ()
#14 0x000000000053b83f in PyEval_EvalFrameEx ()
#15 0x000000000053fc97 in ?? ()
#16 0x000000000053b83f in PyEval_EvalFrameEx ()
#17 0x000000000053fc97 in ?? ()
#18 0x000000000053b83f in PyEval_EvalFrameEx ()
#19 0x0000000000540b0b in PyEval_EvalCodeEx ()
#20 0x00000000004ec2e3 in ?? ()
#21 0x00000000005c20e7 in PyObject_Call ()
#22 0x00000000004fbfce in ?? ()
#23 0x00000000005c20e7 in PyObject_Call ()
#24 0x0000000000574db6 in ?? ()
#25 0x00000000005c20e7 in PyObject_Call ()
#26 0x000000000053b656 in PyEval_EvalFrameEx ()
#27 0x0000000000540b0b in PyEval_EvalCodeEx ()
#28 0x00000000004ec2e3 in ?? ()
#29 0x00000000005c20e7 in PyObject_Call ()
#30 0x00000000004fbfce in ?? ()
#31 0x00000000005c20e7 in PyObject_Call ()
#32 0x0000000000574db6 in ?? ()
#33 0x00000000005c20e7 in PyObject_Call ()
#34 0x000000000053b656 in PyEval_EvalFrameEx ()
#35 0x0000000000540b0b in PyEval_EvalCodeEx ()
#36 0x00000000004ec2e3 in ?? ()
#37 0x00000000005c20e7 in PyObject_Call ()
#38 0x00000000004fbfce in ?? ()
#39 0x00000000005c20e7 in PyObject_Call ()
#40 0x0000000000584d46 in ?? ()
#41 0x00000000005373b6 in PyEval_EvalFrameEx ()
#42 0x00000000005401ef in ?? ()
#43 0x000000000053bc93 in PyEval_EvalFrameEx ()
#44 0x000000000054124a in PyEval_EvalCodeEx ()
#45 0x00000000004ec3f7 in ?? ()
#46 0x00000000005c20e7 in PyObject_Call ()
#47 0x0000000000538cab in PyEval_EvalFrameEx ()
#48 0x000000000053b294 in PyEval_EvalFrameEx ()
#49 0x000000000053b294 in PyEval_EvalFrameEx ()
#50 0x000000000053b294 in PyEval_EvalFrameEx ()
#51 0x0000000000540b0b in PyEval_EvalCodeEx ()
#52 0x00000000004ec2e3 in ?? ()
#53 0x00000000005c20e7 in PyObject_Call ()
---Type <return> to continue, or q <return> to quit---c
#54 0x00000000004fbfce in ?? ()
#55 0x00000000005c20e7 in PyObject_Call ()
#56 0x0000000000574c19 in ?? ()
#57 0x000000000057f58c in ?? ()
#58 0x00000000005c20e7 in PyObject_Call ()
#59 0x000000000053b656 in PyEval_EvalFrameEx ()
#60 0x000000000053b294 in PyEval_EvalFrameEx ()
#61 0x000000000053b294 in PyEval_EvalFrameEx ()
#62 0x000000000053b294 in PyEval_EvalFrameEx ()
#63 0x0000000000540b0b in PyEval_EvalCodeEx ()
#64 0x00000000004ec2e3 in ?? ()
#65 0x00000000005c20e7 in PyObject_Call ()
#66 0x00000000004fbfce in ?? ()
#67 0x00000000005c20e7 in PyObject_Call ()
#68 0x0000000000574c19 in ?? ()
#69 0x000000000057f58c in ?? ()
#70 0x00000000005c20e7 in PyObject_Call ()
#71 0x000000000053b656 in PyEval_EvalFrameEx ()
#72 0x0000000000540b0b in PyEval_EvalCodeEx ()
#73 0x00000000004ec2e3 in ?? ()
#74 0x00000000005c20e7 in PyObject_Call ()
#75 0x00000000004fbfce in ?? ()
#76 0x00000000005c20e7 in PyObject_Call ()
#77 0x000000000057508f in ?? ()
#78 0x00000000005c1f43 in PyObject_GetIter ()
#79 0x0000000000542e9a in ?? ()
#80 0x00000000004ea10f in PyCFunction_Call ()
#81 0x0000000000536d94 in PyEval_EvalFrameEx ()
#82 0x000000000053b294 in PyEval_EvalFrameEx ()
#83 0x000000000053fc97 in ?? ()
#84 0x000000000053b83f in PyEval_EvalFrameEx ()
#85 0x000000000053fc97 in ?? ()
#86 0x000000000053b83f in PyEval_EvalFrameEx ()
#87 0x000000000053fc97 in ?? ()
#88 0x00000000005409bf in PyEval_EvalCode ()
#89 0x000000000060cb42 in ?? ()
#90 0x000000000060efea in PyRun_FileExFlags ()
#91 0x000000000060f7dc in PyRun_SimpleFileExFlags ()
#92 0x0000000000640256 in Py_Main ()
#93 0x00000000004d0001 in main ()
(gdb)
(gdb)
(gdb)
(gdb)
(gdb)

```

**Comment by shaibagon** [comment]
facing same issue here. te core dumps I get are not very informative:
```
(gdb) bt
#0  0x00007f93dc65afb7 in ?? ()
#1  0x0000000000000000 in ?? ()
```

Is there any solution to the problem?
